File: impls/zig/step1_read_print.zig
const reader = @import("reader.zig");
const printer = @import("printer.zig");
const getline = @import("readline.zig").getline;

const Allocator = @import("std").heap.c_allocator;

const MalType = @import("types.zig").MalType;
const get_error_data = @import("error.zig").get_error_data;
const stdout_file = @import("std").io.getStdOut();

fn READ(a: []const u8) !*MalType {
    var read = try reader.read_str(a);
    return reader.read_form(&read);
}

fn EVAL(a: *MalType) *MalType {
    a.incref();
    return a;
}

fn PRINT(mal: MalType) !void {
    try printer.one_stdout(mal);
    try stdout_file.writeAll("\n");
}

fn rep(input: []const u8) !void {
    const read_input = try READ(input);
    defer read_input.decref();
    const eval_input = EVAL(read_input);
    defer eval_input.decref();
    try PRINT(eval_input.*);
}

pub fn main() !void {
    while(try getline("user> ")) |line| {
        defer Allocator.free(line);
        rep(line) catch |err| {
            try stdout_file.writeAll("Error: ");
            try stdout_file.writeAll(@errorName(err));
            try stdout_file.writeAll("\n");
            if(get_error_data()) |mal| {
                defer mal.decref();
                try stdout_file.writeAll("MAL error object is: ");
                try PRINT(mal.*);
            }
        };
    }
}

File: impls/haskell/step1_read_print.hs
import Control.Monad.Except (runExceptT)

import Readline (addHistory, readline, load_history)
import Types
import Reader (read_str)
import Printer (_pr_str)

-- read

mal_read :: String -> IOThrows MalVal
mal_read = read_str

-- eval

eval :: MalVal -> MalVal
eval = id

-- print

mal_print :: MalVal -> IO String
mal_print = _pr_str True

-- repl

repl_loop :: IO ()
repl_loop = do
    line <- readline "user> "
    case line of
        Nothing -> return ()
        Just "" -> repl_loop
        Just str -> do
            addHistory str
            res <- runExceptT $ eval <$> mal_read str
            out <- case res of
                Left mv -> (++) "Error: " <$> mal_print mv
                Right val -> mal_print val
            putStrLn out
            repl_loop

main :: IO ()
main = do
    load_history
    repl_loop
